impl/memory/mutation_logs.go

```diff
@@ -0,0 +1,132 @@
+	return make(MutationLog)
+}
+
+// MutationLog is a fake implementation of keyserver.
+type batch struct {
+	time time.Time
+	msgs []*mutator.LogMessage
+}
+
+// MutationLog is a fake, in-memory implementation of keyserver.MutationLogs
+type MutationLog map[int64][]batch // Map of logID to Slice of LogMessages
+
+// AddLogs to the mutation database.
+func (m MutationLog) AddLogs(ctx context.Context, directoryID string, logIDs ...int64) error {
+	for _, logID := range logIDs {
+		m[logID] = make([]batch, 0)
+	}
+	return nil
+}
+
+// AddLogs to the mutation database.
+func (m MutationLog) ListLogs(ctx context.Context, directoryID string, writable bool) ([]int64, error) {
+	logIDs := []int64{}
+	for logID := range m {
+		logIDs = append(logIDs, logID)
+	}
+	sort.Slice(logIDs, func(a, b int) bool { return logIDs[a] < logIDs[b] })
+	return logIDs, nil
+}
+
+// Send a batch of mutations to a random log.
+func (m MutationLog) Send(ctx context.Context, dirID string, mutation ...*pb.EntryUpdate) (int64, time.Time, error) {
+	// Select a random logID
+	var logID int64
+	for i := range m {
+		logID = i
+		break
+	}
+	ts := time.Now()
+	// Remove the PII contained in EntryUpdate. Only save the merkle tree bits.
+	entries := make([]*pb.SignedEntry, 0, len(mutation))
+	for _, i := range mutation {
+		entries = append(entries, i.Mutation)
+	}
+	m.SendAt(logID, ts, entries)
+	return logID, ts, nil
+}
+
+// SendAt sends a batch of mutations with a given timestamp.
+func (m MutationLog) SendAt(logID int64, ts time.Time, entries []*pb.SignedEntry) {
+	logShard := m[logID]
+	if len(logShard) > 0 && logShard[len(logShard)-1].time.After(ts) {
+		panic(fmt.Sprintf("inserting mutation entry %v out of order", ts))
+	}
+
+	// Convert []SignedEntry into []LogMessage for storage.
+	msgs := make([]*mutator.LogMessage, 0, len(entries))
+	for _, e := range entries {
+		msgs = append(msgs, &mutator.LogMessage{ID: ts, Mutation: e})
+	}
+	m[logID] = append(logShard, batch{time: ts, msgs: msgs})
+}
+
+// ReadLog returns mutations between low, and high. Always returns complete batches.
+func (m MutationLog) ReadLog(ctx context.Context, dirID string,
+	logID int64, low, high time.Time, batchSize int32) ([]*mutator.LogMessage, error) {
+	logShard := m[logID]
+	if len(logShard) == 0 || batchSize == 0 {
+		return nil, nil
+	}
+	start := sort.Search(len(logShard), func(i int) bool { return !logShard[i].time.Before(low) })
+	end := sort.Search(len(logShard), func(i int) bool { return !logShard[i].time.Before(high) })
+	// If the search is unsuccessful, i will be equal to len(logShard).
+	if start == len(logShard) {
+		return nil, nil
+	}
+	out := make([]*mutator.LogMessage, 0, batchSize)
+	count := int32(0)
+	for i := start; i < end; i++ {
+		out = append(out, logShard[i].msgs...)
+		count += int32(len(logShard[i].msgs))
+		if count >= batchSize {
+			break
+		}
+	}
+	return out, nil
+}
+
+// HighWatermark returns the highest timestamp batchSize items beyond start.
```

Comment:
I feel like there should be some elaboration around `batchSize`. Is it necessarily exactly `batchSize` items?